src/TableFunctions/TableFunctionLoop.cpp

```diff
@@ -0,0 +1,156 @@
+
+                size_t dot_pos = id_name.find('.');
+                if (id_name.find('.', dot_pos + 1) != String::npos)
+                    throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "There are more than one dot");
+                if (dot_pos != String::npos)
+                {
+                    loop_database_name = id_name.substr(0, dot_pos);
+                    loop_table_name = id_name.substr(dot_pos + 1);
+                }
+                else
+                {
+                    loop_table_name = id_name;
+                }
+            }
+            else if (const auto * func = args[0]->as<ASTFunction>())
+            {
+                inner_table_function_ast = args[0];
+            }
+            else
+            {
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Expected identifier or function for argument 1 of function 'loop', got {}", args[0]->getID());
+            }
+        }
+            // loop(database, table)
+        else if (args.size() == 2)
+        {
+            args[0] = evaluateConstantExpressionForDatabaseName(args[0], context);
+            args[1] = evaluateConstantExpressionOrIdentifierAsLiteral(args[1], context);
+
+            loop_database_name = checkAndGetLiteralArgument<String>(args[0], "database");
+            loop_table_name = checkAndGetLiteralArgument<String>(args[1], "table");
+        }
+        else
+        {
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Table function 'loop' must have 1 or 2 arguments.");
+        }
+    }
+
+    ColumnsDescription TableFunctionLoop::getActualTableStructure(ContextPtr /*context*/, bool /*is_insert_query*/) const
+    {
+        return ColumnsDescription();
+    }
+
+    StoragePtr TableFunctionLoop::executeImpl(
+            const ASTPtr & /*ast_function*/,
+            ContextPtr context,
+            const std::string & table_name,
+            ColumnsDescription cached_columns,
+            bool is_insert_query) const
+    {
+        StoragePtr storage;
+        if (!loop_table_name.empty())
+        {
+            String database_name = loop_database_name;
+            if (database_name.empty())
+                database_name = context->getCurrentDatabase();
+
+            auto database = DatabaseCatalog::instance().getDatabase(database_name);
+            storage = database->tryGetTable(loop_table_name, context);
+            if (!storage)
+                throw Exception(ErrorCodes::UNKNOWN_TABLE, "Table '{}' not found in database '{}'", loop_table_name, database_name);
+        }
+
+        else
+        {
+            auto inner_table_function = TableFunctionFactory::instance().get(inner_table_function_ast, context);
+            storage = inner_table_function->execute(
+                    inner_table_function_ast,
+                    context,
+                    table_name,
+                    std::move(cached_columns),
+                    is_insert_query);
```

Comment:
The 5th argument of `ITableFunction::execute` is `bool use_global_context` and not `is_insert_query`, see https://github.com/ClickHouse/ClickHouse/blob/3b2a325327b63ef25d5546d69f1132fcf1788d7a/src/TableFunctions/ITableFunction.h#L85

Reply:
Thanks, fixed in https://github.com/ClickHouse/ClickHouse/pull/78707